File: preview/fa_icon.py
"""Parse Quicker fa: icon specs (same rules as preview/static/fa-icon.js)."""

from __future__ import annotations

import re
from dataclasses import dataclass

FA_STYLES: dict[str, str] = {
    "Light": "fal",
    "Solid": "fas",
    "Regular": "far",
    "Brands": "fab",
}

_PASCAL_KEBAB = re.compile(r"([a-z0-9])([A-Z])")


@dataclass(frozen=True)
class FaIconMeta:
    prefix: str
    icon_class: str
    color: str | None


def is_http_icon_url(icon_url: str | None) -> bool:
    if not icon_url:
        return False
    s = icon_url.strip().lower()
    return s.startswith(("http://", "https://", "//"))


def normalize_http_icon_url(icon_url: str | None) -> str | None:
    if not is_http_icon_url(icon_url):
        return None
    s = icon_url.strip()
    return f"https:{s}" if s.startswith("//") else s


def parse_quicker_fa_spec(spec: str | None) -> FaIconMeta | None:
    if not spec:
        return None
    raw = spec.strip()
    if not raw.lower().startswith("fa:"):
        return None

    body = raw[3:].strip()
    if not body:
        return None

    color: str | None = None
    hash_idx = body.find("#")
    if hash_idx >= 0:
        color = body[hash_idx:].strip() or None
        body = body[:hash_idx].rstrip(":").strip()
    else:
        parts = body.split(":")
        if len(parts) > 1:
            tail = parts[-1].strip()
            if tail and "_" not in tail and not tail[0].isupper():
                body = ":".join(parts[:-1]).strip()

    underscore = body.find("_")
    if underscore <= 0:
        return None

    style_key = body[:underscore]
    glyph = body[underscore + 1 :]
    if not glyph:
        return None

    prefix = FA_STYLES.get(style_key, "fal")
    icon_class = f"fa-{_pascal_to_kebab(glyph)}"
    return FaIconMeta(prefix=prefix, icon_class=icon_class, color=color)


def _pascal_to_kebab(name: str) -> str:
    text = name.replace("_", "-")
    return _PASCAL_KEBAB.sub(r"\1-\2", text).lower()

File: preview/web/src/lib/faIcon.ts
const FA_STYLES: Record<string, string> = {
  Light: "fal",
  Solid: "fas",
  Regular: "far",
  Brands: "fab",
};

export interface FaIconMeta {
  prefix: string;
  iconClass: string;
  color: string | null;
}

export function parseQuickerFaSpec(spec: string | null | undefined): FaIconMeta | null {
  if (!spec) return null;
  const raw = spec.trim();
  if (!raw.toLowerCase().startsWith("fa:")) return null;

  let body = raw.slice(3).trim();
  if (!body) return null;

  let color: string | null = null;
  const hashIdx = body.indexOf("#");
  if (hashIdx >= 0) {
    color = body.slice(hashIdx).trim();
    body = body.slice(0, hashIdx).replace(/:+$/, "").trim();
  } else {
    const parts = body.split(":");
    if (parts.length > 1) {
      const tail = parts[parts.length - 1]?.trim() ?? "";
      if (tail && !tail.includes("_") && !/^[A-Z]/.test(tail)) {
        body = parts.slice(0, -1).join(":").trim();
      }
    }
  }

  const underscore = body.indexOf("_");
  if (underscore <= 0) return null;

  const styleKey = body.slice(0, underscore);
  const glyph = body.slice(underscore + 1);
  if (!glyph) return null;

  const prefix = FA_STYLES[styleKey] ?? "fal";
  const iconClass = `fa-${pascalToKebab(glyph)}`;
  return { prefix, iconClass, color };
}

function pascalToKebab(name: string): string {
  return name
    .replace(/_/g, "-")
    .replace(/([a-z0-9])([A-Z])/g, "$1-$2")
    .toLowerCase();
}

export function isHttpIconUrl(iconUrl: string | null | undefined): boolean {
  if (!iconUrl) return false;
  const s = iconUrl.trim().toLowerCase();
  return s.startsWith("http://") || s.startsWith("https://") || s.startsWith("//");
}

export function normalizeHttpIconUrl(iconUrl: string | null | undefined): string | null {
  if (!isHttpIconUrl(iconUrl)) return null;
  const s = iconUrl!.trim();
  return s.startsWith("//") ? `https:${s}` : s;
}
